src/transformers/models/univnet/modeling_univnet.py

```diff
@@ -0,0 +1,677 @@
+        stride: int = 8,
+        dilations: Union[Tuple[int], List[int]] = [1, 3, 9, 27],
+        lvc_hop_size: int = 256,
+        resnet_leaky_relu_slope: float = 0.2,
+        cond_channels: int = 100,
+        kernel_predictor_num_blocks: int = 3,
+        kernel_predictor_hidden_channels: int = 64,
+        kernel_predictor_conv_size: int = 3,
+        kernel_predictor_dropout: float = 0.0,
+        kernel_predictor_leaky_relu_slope: float = 0.2,
+    ):
+        super().__init__()
+        self.cond_hop_length = lvc_hop_size
+        self.leaky_relu_slope = resnet_leaky_relu_slope
+
+        self.convt_pre = nn.ConvTranspose1d(
+            hidden_channels,
+            hidden_channels,
+            2 * stride,
+            stride=stride,
+            padding=stride // 2 + stride % 2,
+            output_padding=stride % 2,
+        )
+
+        self.kernel_predictor = UnivNetKernelPredictor(
+            conv_layers=len(dilations),
+            conv_in_channels=hidden_channels,
+            conv_out_channels=2 * hidden_channels,
+            conv_kernel_size=kernel_size,
+            num_blocks=kernel_predictor_num_blocks,
+            resnet_in_channels=cond_channels,
+            resnet_hidden_channels=kernel_predictor_hidden_channels,
+            resnet_kernel_size=kernel_predictor_conv_size,
+            dropout=kernel_predictor_dropout,
+            leaky_relu_slope=kernel_predictor_leaky_relu_slope,
+        )
+
+        self.resblocks = nn.ModuleList(
+            [
+                UnivNetLVCResidualBlock(
+                    hidden_channels=hidden_channels,
+                    kernel_size=kernel_size,
+                    dilation=dilations[i],
+                    leaky_relu_slope=resnet_leaky_relu_slope,
+                )
+                for i in range(len(dilations))
+            ]
+        )
+
+    def forward(self, hidden_states: torch.FloatTensor, spectrogram: torch.FloatTensor):
+        # hidden_states: (batch_size, hidden_channels, seq_length)
+        # spectrogram: (batch_size, cond_channels, cond_length)
+        hidden_states = nn.functional.leaky_relu(hidden_states, self.leaky_relu_slope)
+        hidden_states = self.convt_pre(hidden_states)
+
+        kernels, biases = self.kernel_predictor(spectrogram)
+
+        for i, resblock in enumerate(self.resblocks):
+            kernel = kernels[:, i, :, :, :, :]
+            bias = biases[:, i, :, :]
+            hidden_states = resblock(hidden_states, kernel, bias, hop_size=self.cond_hop_length)
+
+        return hidden_states
+
+    def apply_weight_norm(self):
+        nn.utils.weight_norm(self.convt_pre)
+        self.kernel_predictor.apply_weight_norm()
+        for layer in self.resblocks:
+            layer.apply_weight_norm()
+
+    def remove_weight_norm(self):
+        nn.utils.remove_weight_norm(self.convt_pre)
+        self.kernel_predictor.remove_weight_norm()
+        for layer in self.resblocks:
+            layer.remove_weight_norm()
```

Comment:
same comment here, why is there a need to `remove` the weight norm?

Reply:
For training, the model with weight norm applied is used, and the original checkpoints also contain the model with weight norm applied (e.g. the state dicts have `*.weight_g`s and `*.weight_v`s). However, for inference the model with weight norm removed is used (see https://github.com/maum-ai/univnet/blob/9bb2b54838bb6d7ce767131cc7b8b61198bc7558/model/generator.py#L61-L65). So it's useful to be able to call `apply_weight_norm` in the conversion script to get the model with weight norm applied, transfer over the weights from the original checkpoint, and then call `remove_weight_norm` to shrink the model back down for storage on the hub and inference.